address_to_census_tract.py

```diff
@@ -7,81 +7,194 @@
+    """
+    print(f"Looking up {address}")    
+    url = build_request_url(address)
+
+    r = requests.get(url)
+    if r.status_code != 200:
+        r.raise_for_status()
+    response = r.json()
+    if len(response) == 0:
+        # Try parsing the address to remove unit or suite #
+        edited_address = reformat_address(address)
+        url = build_request_url(edited_address)
+
+        r = requests.get(url)
+        if r.status_code != 200:
+            r.raise_for_status()
+        response = r.json()
+        if len(response) == 0:
+            raise ValueError(f"Could not geocode address «{address}». \
+No result found.")
+
+    return response[0]
+
+def response_to_standard_address(response: object) -> str:
+    """
+    Returns a compiled and standard address from the given HTTP response object.
+    """
+    standard_address = response['delivery_line_1']
 
-    with open(geojson_filename) as file:
-        geojson = json.load(file)
+    if 'delivery_line_2' in response.keys():
+        standard_address += ' ' + response['delivery_line_2']
 
-    return [
-        { "properties": feature["properties"], "shape": shape(feature['geometry']) }
-            for feature in geojson["features"]
-    ]
+    return standard_address + ', ' + response['last_line']
 
+def response_to_latlng(response: object) -> list:
+    """
+    Returns latitude and longitude coordinates from the given HTTP response
+    object.
+    """
+    return [response['metadata']['latitude'], response['metadata']['longitude']]
 
-def latlng_to_polygon(latlng, polygons):
+def latlng_to_polygon(latlng):
     """
-    Find the first polygon in *polygons* which contains the *latlng* and return
-    the polygon, else None.
+    Find the first polygon in *polygons* (loaded from file) which contains the 
+    *latlng* and return the polygon, else None.
     """
+    polygons = load_geojson("data/geojsons/Washington_2016.geojson")
 
     # Ye olde lat/lng vs. lng/lat schism rears its head.
     lat, lng = latlng
     point = Point(lng, lat)
 
     for polygon in polygons:
         if polygon["shape"].contains(point):
-            return polygon
+            return polygon.get("properties", {}).get("GEOID")
 
+    print(f"failed to find tract for {latlng}", file = stderr)
     return None
 
+def reformat_address(address: str) -> str: 
+    """
+    Removes a stray unit/suite number at the front of a given address. Also
+    removes unit/suite numbers followed by pound sign (#). These unit/suite 
+    numbers often cause errors in geocoding.  
+
+    Assumptions:
+    1. A street address contains at least one letter, [a-z] (case insensitive).
+    2. A street address comes before a city, state, and zip code in an address.
+    3. An address won't have more than one comma-separated string ahead of a 
+    true street address.
+
+    The following method should work even if the given address has no commas
+    separating the address components. 
```

Comment:
I'm not comfortable with the accuracy of trying to parse and reformat addresses ourselves. That's part of reason we're using a third-party API. :-) I wonder why SmartyStreets fails on these? Could it be what we're asking for or the way we're doing a request rather than some fundamental inability? Can we consult with SmartyStreets documentation/support here for a better way to handle the problematic addresses?

What other solutions did you try?

Reply:
Your concerns make sense.

Incidentally, switching to the smartstreets-python-sdk library proved more robust for addresses with the Suite or Unit # issue. 

I'm still looking into fixing the lookup for the other problematic address (with unit number out front, plus a comma).

Reply:
I switched to using SmartyStreets' US Extract API for addresses where the normal lookup fails.